README.md
# script.service.site-monitor


changelog.txt
v0.1.2
- Send 0 to arduino serial.

v0.1.1
- Play song after "count" fails.

v0.1.0
- Console can be used.

v0.0.4
- Notifications can be disabled when server is up.
- Check time can go from one minute to sixty.

v0.0.3
- First official revision


addon.py
import os, sys, time

try:
	import xbmc, xbmcgui, xbmcaddon

	__cwd__ = xbmc.translatePath(xbmcaddon.Addon().getAddonInfo('path')).decode('utf-8')
	__resource__ = xbmc.translatePath(os.path.join(__cwd__, 'resources', 'lib')).decode('utf-8')
	sys.path.append(__resource__)
	from media import *
	model = Model(xbmc, xbmcgui, xbmcaddon)
except ImportError:
	__cwd__ = os.path.dirname(os.path.abspath(__file__)).decode('utf-8')
	__resource__ = os.path.join(__cwd__, 'resources', 'lib').decode('utf-8')
	sys.path.append(__resource__)
	from console import *
	model = Model()

line1 = 'Starting site monitor...'
model.notification(line1)
first = True
while not model.abortRequested():
	if first or model.hasChanged():
		model.read()
		notify = model.getNotify()
		check = model.getCheck()
		count = model.getCount()
		song = model.getSong()
		server = model.getServer()
		first = False
	if server == '':
		continue
	for i in range(0, check):
		if model.abortRequested():
			break
		if model.hasChanged():
			break
		if i == 0:
			if notify == True:
				line1 = 'Checking %s ...' % (server)
				model.notification(line1)
			response = os.system('ping -c 1 ' + server)
			if response == 0:
				if notify == True:
					line1 = '%s is up.' % (server)
					model.notification(line1)
				count = model.getCount()
				sleep = model.getSleep(False)
			else:
				line1 = '%s is down.' % (server)
				model.notification(line1)
				count = count - 1
				if count == 0:
					model.play(song)
					count = model.getCount()
				sleep = model.getSleep(True)
		time.sleep(sleep)


default.py
import os, sys, time

try:
	import xbmc, xbmcgui, xbmcaddon

	__cwd__ = xbmc.translatePath(xbmcaddon.Addon().getAddonInfo('path')).decode('utf-8')
	__resource__ = xbmc.translatePath(os.path.join(__cwd__, 'resources', 'lib')).decode('utf-8')
	sys.path.append(__resource__)
	from media import *
	model = Model(xbmc, xbmcgui, xbmcaddon)
except ImportError:
	__cwd__ = os.path.dirname(os.path.abspath(__file__)).decode('utf-8')
	__resource__ = os.path.join(__cwd__, 'resources', 'lib').decode('utf-8')
	sys.path.append(__resource__)
	from console import *
	model = Model()

line1 = 'Starting site monitor...'
model.notification(line1)
first = True
while not model.abortRequested():
	if first or model.hasChanged():
		model.read()
		notify = model.getNotify()
		check = model.getCheck()
		count = model.getCount()
		song = model.getSong()
		server = model.getServer()
		first = False
	if server == '':
		continue
	for i in range(0, check):
		if model.abortRequested():
			break
		if model.hasChanged():
			break
		if i == 0:
			if notify == True:
				line1 = 'Checking %s ...' % (server)
				model.notification(line1)
			response = os.system('ping -c 1 ' + server)
			if response == 0:
				if notify == True:
					line1 = '%s is up.' % (server)
					model.notification(line1)
				sleep = model.getSleep(False)
			else:
				line1 = '%s is down.' % (server)
				model.notification(line1)
				count = count - 1
				if count == 0:
					model.play(song)
					count = model.getCount()
				sleep = model.getSleep(True)
		time.sleep(sleep)


resources/lib/console.py
import signal, sys, serial, time;

__should_exit__ = False

def handler(signum, frame):
	global __should_exit__
	__should_exit__ = True

class Model(object):
	def __init__(self):
		self._arduino = False
		signal.signal(signal.SIGINT, handler)

	def notification(self, line1):
		print line1

	def abortRequested(self):
		return __should_exit__

	def read(self):
		pass

	def hasChanged(self):
		return False

	def getServer(self):
		return 'home.vicenteobregon.com'

	def getCheck(self):
		return 1

	def getCount(self):
		return 3

	def getNotify(self):
		return True

	def getSong(self):
		return '\a'

	def play(self, song):
		if self._arduino == False:
			try:
				ser = serial.Serial('/dev/arduino', 9600)
				time.sleep(2)
				ser.write(b'0')
				ser.close()
				self._arduino = True
			except:
				pass
		print song

	def getSleep(self, error):
		if error:
			return 1
		else:
			self._arduino = False;
			return 60


resources/lib/media.py
import serial, time;

class Model(object):
	def __init__(self, xbmc, xbmcgui, xbmcaddon):
		self._xbmc = xbmc
		self._xbmcgui = xbmcgui
		self._xbmcaddon = xbmcaddon.Addon()
		self._addonname = self._xbmcaddon.getAddonInfo('name')
		self._icon = self._xbmcaddon.getAddonInfo('icon')
		self._time = 5000
		self._arduino = False

	def notification(self, line1):
		self._xbmcgui.Dialog().notification(self._addonname, line1, self._icon, self._time)

	def abortRequested(self):
		return self._xbmc.abortRequested

	def read(self):
		self._notify = (self._xbmcaddon.getSetting('notify') == 'true')
		self._check = int(self._xbmcaddon.getSetting('check'))
		self._count = int(self._xbmcaddon.getSetting('count'))
		self._song = self._xbmcaddon.getSetting('song')
		self._server = self._xbmcaddon.getSetting('server1')

	def hasChanged(self):
		notify = (self._xbmcaddon.getSetting('notify') == 'true')
		if self._notify != notify:
			return True
		check = int(self._xbmcaddon.getSetting('check'))
		if self._check != check:
			return True
		count = int(self._xbmcaddon.getSetting('count'))
		if self._count != count:
			return True
		song = self._xbmcaddon.getSetting('song')
		if self._song != song:
			return True
		server = self._xbmcaddon.getSetting('server1')
		if self._server != server:
			return True
		return False

	def getNotify(self):
		return self._notify;

	def getCheck(self):
		return self._check

	def getCount(self):
		return self._count

	def getSong(self):
		return self._song

	def getServer(self):
		return self._server

	def play(self, song):
		if self._arduino == False:
			try:
				ser = serial.Serial('/dev/arduino', 9600)
				time.sleep(2)
				ser.write(b'0')
				ser.close()
				self._arduino = True
			except:
				pass
		self._xbmc.Player().play(song)

	def getSleep(self, error):
		if not error:
			self._arduino = False
		return 60
